### DISCUSSION TITLE
1322 : 切木棍

### DISCUSSION THREAD

**Discussion by dgsvygd** [CSGrandeur/s-1problem1day1ac#442] [OJ Problem Discussion]
基础的二分答案模板题

可以观察到，如果直接计算答案就会有点无从下手

但是我们可以通过 `O(n)` 的复杂度，去判断一个我们假设的答案是否成立，而且显然答案最终答案只会存在于 1 到 初始木棍最长长度 之间，我们可以考虑从最大到最小去假设最终答案，时间复杂度为 `O(n²)`，显然还需要优化

以样例为例：成立的情况表示为 1，不成立的情况表示为 0

| 假设最终答案为                 | 1    | 2    | 3    | 4    | 5    | 6    | 7    | 8    | 9    |
| ------------------------------ | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- |
| 是否成立 | 0    | 0    | 0    | 1    | 1    | 1    | 1    | 1    | 1    |

我们可以发现，答案是单调的，因此我们可以用二分去优化枚举答案的过程，时间复杂度为 `O(nlogn)`

```cpp
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <vector>
#include <string>
#include <queue>
#include <functional>
#include <map>
#include <set>
#include <cmath>
#include <cstring>
#include <deque>
#include <stack>
using namespace std;
typedef long long ll;
#define pii pair<int, int>
const ll maxn = 2e5 + 10;
const ll inf = 1e17 + 10;
int num[maxn]; // 输入
int n, k;

bool judge(int x) // 用来判断，如果最终切成最长长度为 x，k 刀够不够
{
    ll cnt = 0;
    for(int i=0; i<n && cnt <= k; i++) // 对当前每个木棍，计算要切多少刀，使得最长长度为 x
    {
        cnt += (num[i] - 1) / x;
    }
    return cnt <= k;
}

int main()
{
    
    scanf("%d%d", &n, &k);
    int r = 0;
    for(int i=0; i<n; i++)
    {
        scanf("%d", &num[i]);
        r = num[i] > r ? num[i] : r; // 二分上界，最终答案最大即为当前木棍最长的
    }
    int l = 1;
    while(l < r) // 二分
    {
        int mid = l + (r - l) / 2; // 中间数
        if(judge(mid)) // 判断当前是否成立，如果成立，则上界置为 mid，否则下界置为 mid + 1
            r = mid;
        else
            l = mid + 1;
    }
    printf("%d\n", l);
    

    return 0;
}


```

**Comment by d3vw** [comment]
字词
judge函数里面那句 `cnt += (num[i] - 1) / x;`
在老师给的代码里写成 `cnt += (num[i] -1 + x) / x - 1;`
把我看懵了哈哈

**Reply by dgsvygd** [reply]
噗，可能就是想的不一样罢了，结果都一样的

**Reply by dgsvygd** [reply]
他的意思大概就是，将当前木棍 num[i] 分成若干个 x 长度的木棍，能分多少段：(num[i] - 1 + x) / x，这个是向上取整，就最后一段如果没有到 x，也把他当成 x
知道了能分成多少段，直接 -1 就是要切的刀数
一般我们考虑 x 除以 k 向上取整：(x + k - 1) / k

**Reply by CSGrandeur** [reply]
我也有点忘了之前为什么这么写了，不过按切多少段来思考是合理的。

其实式子化简一下就没了，只是当时没多想。
